Fix spline interpolation boundary handling for modes reflect and wrap

grlee77 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
closes gh-1323
closes gh-8131
closes gh-8465
resolves: scikit-image/scikit-image#1585

resolves part of gh-2640
   - 'reflect' and and the new 'grid-wrap' have correct boundary behavior (the name `grid-wrap` was suggested by @stefanv: https://github.com/scipy/scipy/pull/206#issuecomment-408326019)
   - modes 'constant' and 'nearest' can still be improved in a subsequent PR

@rgommers, @stefanv: this is the second PR related to the NumFOCUS proposal.

#### What does this implement/fix?
<!--Please explain your changes.-->
This PR makes a number of improvements to boundary handling in ndimage. This is related to prior discussions in #206, #2640, and #8465. It is probably easiest to review the commits in order rather than all at once.

This PR fixes boundary handling when mode='reflect' and introduces a new mode, `'wrap-grid'` that behaves according to the description of wrap in the docstring. Prior to this PR, the requested `mode` is only used when wrapping the current coordinate into the original signal range:
https://github.com/scipy/scipy/blob/d866f8342a1c0f1fb8175460e520ea264bca1ca5/scipy/ndimage/src/ni_interpolation.c#L409). 
However, `mirror` mode was then [always used for other coefficients involved in the interpolation](https://github.com/scipy/scipy/blob/d866f8342a1c0f1fb8175460e520ea264bca1ca5/scipy/ndimage/src/ni_interpolation.c#L426-L439). This was done because previously, only mirror boundary conditions had been implemented for spline orders > 1. However as of #8537, we also have boundary conditions implemented for reflect and wrap cases.

For backwards compatibility, I have kept the current `wrap` as-is, but am not sure if we want to keep it or should just replace it with the new behavior. I also added a `grid-mirror` alias for the name `reflect` as suggested in https://github.com/scipy/scipy/pull/206#issuecomment-408326019, but I'm not sure if we really want/need to add that.

#### Additional information
<!--Any additional information you think is important.-->

Prior to this PR, boundary handling for both 'reflect' and 'wrap' were not as expected for any spline order.

The boundary handling for `mode='nearest'` was also improved slightly for order >1, but I intend to revisit this again and see if the remaining error can be avoided in a subsequent PR. mode='constant'` will also be revised in a future PR.

## Updated example based on the one given in https://github.com/scipy/scipy/issues/2640#issuecomment-554873759
The figure in the comment is still a match to current master. The new behavior after this PR is shown below for order=3:

![compare_to_padded_order3](https://user-images.githubusercontent.com/6528957/91190372-81dc5a80-e6c1-11ea-99e7-50e706b8310a.png)


Note that unlike on current master the existing `'reflect'` mode and the new `'grid-wrap'` mode match the case where explicit padding/extension of the signal was performed prior to calling the interpolation.


**TODO:**

- [x] add new test cases

grlee77 commented:
Additional figures for order=1 and order=0

![compare_to_padded_order1](https://user-images.githubusercontent.com/6528957/91191563-f19f1500-e6c2-11ea-9974-96df4eeb570b.png)

![compare_to_padded_order0](https://user-images.githubusercontent.com/6528957/91191574-f4016f00-e6c2-11ea-86b1-7461bf98fa52.png)

grlee77 commented:
The following figure shows dashed lines at the point where wraps and/or reflections occur. As can be seen the interval included for both `reflect` and `wrap-grid` extends by 0.5 outside of the original signal extent on either side. 

![wrap_grid_order3](https://user-images.githubusercontent.com/6528957/91191998-712ce400-e6c3-11ea-8236-ee2c2509f5d4.png)

larsoner commented:
These figures are very informative! Could you add them to an `Examples` section to some appropriate function so that they actually get embedded in the built doc?

stefanv commented:
@grlee77 This is superb work, and you have no idea what great joy it gives me to see these issues finally addressed and clarified!

In terms of constant, do you think we need `grid-constant` as well?

Eventually, we may want to deprecate `reflect` in favor of `grid-mirror`, so that all arguments are consistent.

Great find on the kernels using mirroring internally—I never realized this and, admittedly, I have not thought through the ramifications.

grlee77 commented:
> In terms of constant, do you think we need grid-constant as well?

I think we probably should add it. See #12776 for what I had come up with so far toward that.


> These figures are very informative! Could you add them to an Examples section to some appropriate function so that they actually get embedded in the built doc?

Yes, that is a good idea. I still need to update some docstrings as well.

grlee77 commented:
Another API point to consider: The functions in `ndimage.filters` also accept a mode argument, but only operate on integer coordinates. For those, the mode 'wrap' already operated with the expected `grid-wrap` behavior. Since `wrap` and `grid-wrap` are equivalent for these, should we just allow users to specify either name or should we deprecate one of the two?

Specifically, there is code for boundary handling in the function `NI_InitFilterOffsets`, which gets used by `NI_Correlate`, `NI_MinOrMaxFilter`, `NI_RankFilter` and `NI_GenericFilter`. It is similar to the code in the `map_coordinates` function used by the interpolation routines, but the cases are in-line rather than in a separate function and the coordinates involved are integers rather than doubles.

There is also separate boundary handling code in `NI_ExtendLine` which gets used by `NI_SplineFilter1D`, `NI_Correlate1D`, `NI_UniformFilter1D`, `NI_MinOrMaxFilter1D` and `NI_GenericFilter1D`. The code in this function actually fills in all of the values in the extended line rather than returning a mapped coordinate.

stefanv commented:
I'd say both, to be consistent, perhaps with a note to indicate as much.

grlee77 commented:
I have now added some test cases and added 'grid-wrap' to the interpolation function docstrings.

There are also now [two new figures in the ndimage tutorial illustrating the boundary modes and integer image coordinates](https://21186-1460385-gh.circle-artifacts.com/0/html-scipyorg/tutorial/ndimage.html#id3).

There are some CI failures that look unrelated. One is a fortran-related failure in building quadpack while most others are something going wrong early on in distutils:
```
      File "c:\hostedtoolcache\windows\python\3.6.8\x86\lib\site-packages\numpy\distutils\ccompiler.py", line 90, in <lambda>
        m = lambda self, *args, **kw: func(self, *args, **kw)
    TypeError: CCompiler_spawn() got an unexpected keyword argument 'env'
```

larsoner commented:
Possibly fixed by https://github.com/scipy/scipy/pull/12795 ?

sethtroisi commented on doc/source/tutorial/examples/plot_interp_grid.py:
```diff
@@ -0,0 +1,63 @@
+from matplotlib import ticker
+
+
+fig, axes = plt.subplots(1, 2, figsize=(11, 5))
+orders = [2, 3]
```
from the zip below I believe axes should have same size as order so consider
```
orders = [2, 3]
fig, axes = plt.subplots(1, len(orders), figsize=(11,5))
```

grlee77 replied:
thanks, I have made the suggested change

grlee77 commented on doc/source/tutorial/ndimage.rst:
```diff
@@ -770,7 +770,26 @@ described in :ref:`ndimage-filter-functions` for the multidimensional
+is used. The behavior of all modes, including at non-integer locations is
+illustrated below. Note that the modes `reflect` (aka `grid-mirror`) and
+`grid-wrap` involve symmetry or repetition about a point that is half way
+between image samples (dashed vertical lines). On the contray, modes `mirror`
```
Thanks, I this does sound better to me.

grlee77 commented on scipy/ndimage/tests/test_interpolation.py:
```diff
@@ -51,6 +52,32 @@ def shift(x):
+                                           mode=mode)
+
+        if mode in ['reflect', 'grid-mirror'] and order > 1:
+            numpy.testing.assert_allclose(y, expected, rtol=1e-4, atol=1e-13)
```
I looked into this relatively high error for mode='reflect' a bit more. It actually is not a problem if the original signal is > 10 samples in length or so. This must be some limitation of the implementation of that boundary mode for short signal lengths.

If we change the test signal above to something like `data = numpy.arange(-3, 8, dtype=float)`, the tests will pass with `rtol=1e-7` in all cases.

grlee77 replied:
I think the inaccuracy is introduced here where the summation is only over a small number of elements when the length, `n`, is short. For longer `n`, `z_i = z**n` becomes close to zero.

https://github.com/scipy/scipy/blob/30e9c92eb89cfc92d6988fb77402475c24f90758/scipy/ndimage/src/ni_splines.c#L218-L222

To fix this for short signals, we would have to operate on an extended signal instead of the original signal length.

For long signals, this loop can probably be truncated at something `min(n, 32)` samples. This is because `z_i` becomes extremely over time in the loop (all filter poles, z, are < 1.0).

The largest pole, `z`, for any of the implemented splines is ~0.43 for the 5th order spline.  `0.43 ** 32` = 1.8e-12, so there is not much point in going to many additional terms.

rgommers replied:
I think this is not worth the effort to improve to be honest. Signal that short are uncommon and will have bigger issues than slightly large `rtol` for signal extrapolation.

grlee77 commented:
I updated the docs in commit 872ac5c based on the feedback in gh-12773 (this PR is prior to the introduction of `grid-constant`, so that is the only one missing in the doc artifact links below):

Docstrings for interpolation functions now link to the ndimage tutorial figures (see modes section): e.g. [affine_transform docstring](https://21441-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.ndimage.affine_transform.html#scipy.ndimage.affine_transform)

Updated mode table and description in the [ndimage filter tutorial](https://21441-1460385-gh.circle-artifacts.com/0/html-scipyorg/tutorial/ndimage.html#filter-functions)

filter function docstrings now mention grid-wrap and grid-mirror as synonyms:  e.g. [convolve docstring](https://21441-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.ndimage.convolve.html#scipy.ndimage.convolve)

larsoner reviewed (approved):
Changes look good to me and seem well tested. @stefanv did you want to look?

stefanv commented:
Fantastic job, thanks @grlee77!  I would like to see https://github.com/scipy/scipy/pull/12776 merged also.

larsoner commented:
Thanks @grlee77 !

shoyer commented:
I am so excited to see this challenging fix complete! Thank you @grlee77 and everyone who helped out!

rgommers commented:
This is awesome, thanks again @grlee77! Would you be able to summarize all the improvements for `ndimage` in the release notes? https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.6.0

grlee77 commented:
A summary of ndimage changes was added at: https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.6.0#scipyndimage-improvements
